### ISSUE TITLE
[capture] External memory in page guard option not spec compliant

### ISSUE THREAD

**Issue by per-mathisen-arm** [LunarG/gfxreconstruct#1133]
I was poking around at the external memory page guard option wondering why it was Windows-only, as several Linux desktop drivers now seem to have the required extension support. Is there a particular reason other than missing extension support that this has been made Windows-only?

The background to why I'm looking at this is that we've been seeing issues with running out of memory and synchronization issues between real and shadow memory, and I was hoping the external memory option could solve those.

To the problem: Binding buffers and images to imported external memory requires the use of VkExternalMemoryBufferCreateInfo and VkExternalMemoryImageCreateInfo extension structs to vkBindBufferMemory and vkBindImageMemory, respectively. See VUID-vkBindBufferMemory-memory-02985 and VUID-vkBindImageMemory-memory-02989. This seems quite tricky, since at the time of creation you don't know to which memory the traced application intends to bind these objects. That is, if it is meant to be used with host visible memory or some other memory which cannot be replaced with external memory. Also, to make matters worse, you would need to verify ahead of time that the image format supports external memory (see VUID-VkImageCreateInfo-pNext-00990).

I don't think marking an image/buffer as using external memory if it isn't going to be used with external memory is a good idea, although I don't see the spec explicitly prohibiting it. I assume doing this is unintended and there might be side effects. The only other option I can think of (aside from just ignoring the whole problem and hoping it doesn't result in anything more than validation layer error spam), is to recreate buffers and images bound to external memory at bind time and remap all their handles... which seems very ugly.

And finally there is the spec for VkAllocateMemory that seemingly explicitly prohibits substituting the memory in the first place: "successfully returned allocation must use the requested parameters — no substitution is permitted by the implementation." (https://registry.khronos.org/vulkan/specs/1.3-extensions/man/html/vkAllocateMemory.html) Although this is a problem also for the shadow memory implementation.

**Comment by panos-lunarg** [comment]
I am not sure why this is a Windows-only option. Looking through the code I think it can be supported for Linux as well.

I see that the external memory option is not using the memory protection method implemented for shadow memory but it is using a different method. I am not sure why it was implemented this way but, as far as I know and please correct me if I'm wrong, there is no equivalent method for Linux/Android. So the Linux/Android version will have to use the memory protection method as well in this case, although it will have to protect the memory only for writes.

You mention you are having sync issues between shadow and mapped memory. Can you elaborate a bit more on that? Do you believe there is a bug in GFXR or are you dealing with an inherent limitation of the shadow memory method?
Perhaps if the external memory method is implemented for Linux it will also suffer from those sync issues as it will use the memory protection method.

**Comment by per-mathisen-arm** [comment]
It is possible for an application to have both CPU and GPU writing to the same memory, provided it does proper synchronization of these writes. But when you use shadow memory, the GPU will write to the real memory while the CPU will write to the shadow memory instead of working together on the same memory. I do not see a way to completely fix this without ruining tracing performance, which is why I was looking into using external memory on Linux/Android instead, which should not have this problem.

**Comment by panos-lunarg** [comment]
Right, so you basically want the cpu writes to be reflected "instantly" to the mapped memory and not wait until queue submit? Yes this will practically ruin performance. Have you tried modifying the code to achieve it though? Perhaps it doesn't ruin it that much in your case, and if so maybe it would make sense to have this as option controlled by an env variable.

I can try to write a "quick" patch to enable the external option for Linux/Android, although I can not guarantee anything for spec compliance as at the moment I don't have any suggestions on how to fix the problem you have discovered.

**Comment by per-mathisen-arm** [comment]
Yes, and I am mostly interested in a general solution here. If you make a patch, I'd be happy to test it out.

**Comment by panos-lunarg** [comment]
Something that crossed my mind. Is your application going to be "capture-able"?
User accesses to host visible memory detected with both methods (external memory and shadow memory) are being written to the capture file when either one of the following functions are called:
1. `vkUnmapMemory`
2. `vkFlushMappedMemoryRanges`
3. `vkQueueSubmit(2)`

(Also in the shadow memory method apart from being written to the file, the changes are also copied to the actual mapped memory)

So with external memory, while capturing the application, the memory changes will be directly visible to the gpu but will they get written to the file in order to be replayed as expected?

**Comment by panos-lunarg** [comment]
Pushed an experimental patch at https://github.com/panos-lunarg/gfxreconstruct/tree/panos_use_write_watch_linux
AMD and Intel drivers (both on mesa) I tried it on didn't seem to like it. Proprietary nvidia driver didn't seem to complain about it but I have only tested it with vkcube so far. Give it a spin and let me know

**Comment by panos-lunarg** [comment]
After some experimentation I think there is a reason this option is not available on Linux.

Initially I thought the crashes (queue submissions return `VK_DEVICE_LOST`) on amd and intel were because of the drivers having poorly implemented/bug `VK_EXT_external_memory_host`. But when I tried giving full read+write access on the mapped memory the crash went away.
I guess this means that the kernel driver or GPU are doing something with the memory and this triggers the device lost.
On windows the memory writes are detected with a different mechanism so there is no problem there.

If [SOFT-DIRTY PTEs](https://www.kernel.org/doc/Documentation/vm/soft-dirty.txt) worked on Android then I think we would have an alternative solution

**Comment by per-mathisen-arm** [comment]
Thanks. I did some testing and it seems to work on Nvidia linux desktop. I think the patch is still worthwhile, even though it won't work on Intel/AMD drivers. It is possible that the failures on Intel/AMD drivers might be related to the incomplete implementation of the external memory support, as noted in the original ticket description above, and that the drivers might hold off on poking into the memory and dying if the objects had been properly marked as coming from external memory. But this is just speculation.
